database/ydb/ydb.go

```diff
@@ -0,0 +1,311 @@
+		return nil, err
+	}
+	return db, nil
+}
+
+func (db *YDB) Open(dsn string) (database.Driver, error) {
+	purl, err := url.Parse(dsn)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(purl.Path) == 0 {
+		return nil, ErrNoDatabaseName
+	}
+
+	pquery, err := url.ParseQuery(purl.RawQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	case pquery.Has(queryParamUseGRPCS):
+		purl.Scheme = "grpcs"
+	default:
+		purl.Scheme = "grpc"
+	}
+
+	purl = migrate.FilterCustomQuery(purl)
+
+	credentials := db.parseCredentialsOptions(purl, pquery)
+	tlsOptions, err := db.parseTLSOptions(purl, pquery)
+	if err != nil {
+		return nil, err
+	}
+
+	driver, err := ydb.Open(context.TODO(), purl.String(), append(tlsOptions, credentials)...)
+	if err != nil {
+		return nil, err
+	}
+
+	px, err := WithInstance(driver, &Config{
+		MigrationsTable: pquery.Get(queryParamMigrationsTable),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return px, nil
+}
+
+func (db *YDB) parseCredentialsOptions(url *url.URL, query url.Values) (credentials ydb.Option) {
+	switch {
+	case query.Has(queryParamAuthToken):
+		credentials = ydb.WithAccessTokenCredentials(query.Get(queryParamAuthToken))
+	case url.User != nil:
+		user := url.User.Username()
+		password, _ := url.User.Password()
+		credentials = ydb.WithStaticCredentials(user, password)
+	default:
+		credentials = ydb.WithAnonymousCredentials()
+	}
+	url.User = nil
+	return credentials
+}
+
+func (db *YDB) parseTLSOptions(_ *url.URL, query url.Values) (options []ydb.Option, err error) {
+	if query.Has(queryParamTLSCertificateAuthorities) {
+		options = append(options, ydb.WithCertificatesFromFile(query.Get(queryParamTLSCertificateAuthorities)))
+	}
+	if query.Has(queryParamTLSInsecureSkipVerify) {
+		options = append(options, ydb.WithTLSSInsecureSkipVerify())
+	}
+	if query.Has(queryParamTLSMinVersion) {
+		switch query.Get(queryParamTLSMinVersion) {
+		case "1.0":
+			options = append(options, ydb.WithMinTLSVersion(tls.VersionTLS10))
+		case "1.1":
+			options = append(options, ydb.WithMinTLSVersion(tls.VersionTLS11))
+		case "1.2":
+			options = append(options, ydb.WithMinTLSVersion(tls.VersionTLS12))
+		case "1.3":
+			options = append(options, ydb.WithMinTLSVersion(tls.VersionTLS13))
+		default:
+			return nil, ErrUnsupportedTLSVersion
+		}
+	}
+	return options, nil
+}
+
+func (db *YDB) Close() error {
+	return db.driver.Close(context.TODO())
+}
+
+func (db *YDB) Run(migration io.Reader) error {
+	rawMigrations, err := io.ReadAll(migration)
+	if err != nil {
+		return err
+	}
+
+	res, err := db.driver.Scripting().Execute(context.TODO(), string(rawMigrations), nil)
+	if err != nil {
+		return err
+	}
+	return res.Close()
+}
+
+func (db *YDB) SetVersion(version int, dirty bool) error {
+	deleteVersionQuery := fmt.Sprintf(`
+		DELETE FROM %s 
+	`, db.config.MigrationsTable)
+
+	insertVersionQuery := fmt.Sprintf(`
+		INSERT INTO %s (version, dirty, created) VALUES (%d, %t, CurrentUtcTimestamp())
```

Comment:
We can use a query with parameters here to eliminate compilation time from migration progress. I am not sure if it is meaningful, though, it is okay to do it with a `Sprintf`.

Reply:
I guess the difference would be in µs and this is a script which will be run manually, so I think `Sprintf` would be ok:)